### DISCUSSION TITLE
Container Role Configuration to Optimize Resource Utilization and Reduce Memory Usage

### DISCUSSION THREAD

**Discussion by okaufmann** [serversideup/docker-php#215] [Q&A] [answered]
Hi guys, 

I am utilizing the `serversideup/php` image to orchestrate a Laravel application across three distinct containers:

- **Webapp Container**: Hosts nginx and PHP-FPM, exposing a designated port for application accessibility.
- **Queue Container**: Dedicated to running Horizon.
- **Scheduler Container**: Executes the `php artisan schedule:work` command.

This architecture facilitates individual scaling of each component on Kubernetes.

However, I've observed that nginx and PHP-FPM are initialized in every container, which significantly increases the Memory usage, especially when managing numerous Laravel instances.

To mitigate this, I've resorted to creating two unique images:
1. A webapp image derived from `serversideup/php:8.2-fpm-nginx`.
2. A shared image for the queue and scheduler containers based on `serversideup/php:8.2-cli`.

This workaround, albeit functional, introduces a complex initial build configuration and heightens the potential for failure due to the need to manage multiple images.

In the past, I've employed a container role concept. Despite the image having the capability to host the webapp, assigning a role like `queue` via environment variable prevented the initialization of nginx and PHP-FPM, thereby optimizing Memory utilization.

I am inquiring if a similar container role feature could be envisaged for future releases, to streamline resource utilization, reduce Memory usage, and simplify image management.

**Comment by jaydrogers** [comment, marked as answer]
In our use case, we do the same thing. We use `fpm-nginx` multiple times (even for things like Queue and Scheduler). In our case this works fine. An app with a similar structure runs on a server running 2GB of RAM and 1 CPU.

Optimizing servers is very unique to every application. In the case if you do not want to use the method above, you will need to use an `FPM` image only and ensure that your server is only processing PHP requests.

This will require extra work where you'll need to ensure you have a load balancer and/or HTTP server outside of your PHP Containers and you'll need to make sure the static assets are stored in an accessible location for all containers.

Hope this helps!

**Reply by okaufmann** [reply]
I think I have expressed myself inaccurately. My problem is that for the queue the corresponding container also starts NGINX and FPM (with s3) which eats memory. But I want to start only the queue without starting NGINX and FPM. To save memory. We run a large amount of Laravel applications and it is important that the individual applications only use as much memory as necessary.

**Reply by jaydrogers** [reply]
Check the `PHP_PM_*` variables here: https://serversideup.net/open-source/docker-php/docs/reference/environment-variable-specification

You might be able to really scale these numbers down. The biggest one for memory savings for me was setting `PHP_PM_CONTROL` to `ondemand`, which saved a ton of memory (which is the default for the image)

**Reply by okaufmann** [reply]
PHP_PM_CONTROL is already set to ondemand by default:

<img width="813" alt="image" src="https://github.com/serversideup/docker-php/assets/4414498/43488fd5-13ba-41c8-8051-b6fd1b86d3f4">

I started my base image and serversideup/php:8.2-cli with the command sleep 60 to compare the memory usage.

<img width="593" alt="image" src="https://github.com/serversideup/docker-php/assets/4414498/554cc7b0-ddc4-4b42-b817-b6773cc21384">
<img width="593" alt="image" src="https://github.com/serversideup/docker-php/assets/4414498/f535c61b-26a1-455c-a17e-057ba3e96d27">


Seems like nginx+fpm consumes about 20Mb more than the php-cli container. Exactly this 20Mb I would like to save.

**Reply by jaydrogers** [reply]
Your other option may be overriding the entrypoints so it doesn't kick off the S6 processor supervision of both services? https://docs.docker.com/compose/compose-file/compose-file-v3/#entrypoint

**Reply by okaufmann** [reply]
I've been searching so much for configuration values and messing around with the s6 that I completely forgot about the simple stuff.... a bit embarrassing 🤦‍♂️ Thanks for the tip! That has reduced the memory consumption.
